File: techniques/dispersion/python/dispersion.py
"""Measures of dispersion / spread (Reference §1.2).

From-scratch implementations plus numpy / scipy equivalents.

Estimators
----------
- range              : max - min
- variance           : sum((x - mean)^2) / (n - ddof)   (ddof=1 -> sample variance)
- standard deviation : sqrt(variance)
- IQR                : Q3 - Q1
- coefficient of var : sd / mean              (unit-free; see techniques/coefficient-of-variation)
- mean abs deviation : mean(|x - center|)     (center = mean or median)
- median abs dev MAD : median(|x - median|)   (robust; *1.4826 to estimate sigma under normality)
"""
from __future__ import annotations    # stdlib: postpone type-hint evaluation (lets us write int | None)

from typing import Sequence    # stdlib: type hint meaning 'indexable iterable' (list / tuple / array)

import numpy as np    # numerical arrays + linear algebra (np.mean, np.linalg.lstsq, ...)


def value_range(x: Sequence[float]) -> float:
    """Range = max(x) - min(x).  ``x`` is a sample."""
    return max(x) - min(x)


def variance(x: Sequence[float], ddof: int = 1) -> float:
    """Sample variance = sum((x - x_bar)^2) / (n - ddof).

    Parameters
    ----------
    x : sample.
    ddof : "delta degrees of freedom" -- divisor is ``n - ddof``. ``1`` (default) gives
        the unbiased sample variance; ``0`` gives the population variance.
    """
    x = list(x)
    n = len(x)
    if n - ddof <= 0:
        raise ValueError("not enough data for the requested ddof")
    m = sum(x) / n
    return sum((xi - m) ** 2 for xi in x) / (n - ddof)


def stddev(x: Sequence[float], ddof: int = 1) -> float:
    """Standard deviation = sqrt(variance(x, ddof)). See ``variance`` for ``ddof``."""
    return variance(x, ddof) ** 0.5


def quantile(x: Sequence[float], p: float) -> float:
    """Linear-interpolation quantile (numpy's default / Hyndman-Fan type 7).

    Parameters
    ----------
    x : sample.
    p : probability in ``[0, 1]`` (e.g. ``0.5`` for the median, ``0.25`` for Q1).
    """
    s = sorted(x)
    n = len(s)
    if n == 1:
        return float(s[0])
    h = (n - 1) * p
    lo = int(h)
    frac = h - lo
    hi = min(lo + 1, n - 1)
    return s[lo] + frac * (s[hi] - s[lo])


def iqr(x: Sequence[float]) -> float:
    """Interquartile range = Q3 - Q1. Robust spread of the middle 50%. ``x`` is a sample."""
    return quantile(x, 0.75) - quantile(x, 0.25)


def coefficient_of_variation(x: Sequence[float], ddof: int = 1) -> float:
    """CV = SD(x) / mean(x). Unit-free relative spread; undefined when the mean is zero.

    ``ddof`` is passed through to ``stddev``.
    """
    m = sum(x) / len(x)
    if m == 0:
        raise ValueError("CV is undefined when the mean is zero")
    return stddev(x, ddof) / m


def mean_abs_deviation(x: Sequence[float], center: str = "mean") -> float:
    """Mean absolute deviation about ``center`` ("mean" or "median"). ``x`` is a sample."""
    x = list(x)
    c = (sum(x) / len(x)) if center == "mean" else _median(x)
    return sum(abs(xi - c) for xi in x) / len(x)


def median_abs_deviation(x: Sequence[float], scale: float = 1.4826) -> float:
    """Median absolute deviation = median(|x - median(x)|) * scale.

    Parameters
    ----------
    x : sample.
    scale : multiplier; ``1.4826`` makes the MAD a consistent estimator of sigma
        under normality. Use ``1.0`` for the raw MAD.
    """
    x = list(x)
    med = _median(x)
    return _median([abs(xi - med) for xi in x]) * scale


def _median(x):
    """Internal helper -- median without re-importing other modules."""
    s = sorted(x)
    n = len(s)
    return s[n // 2] if n % 2 else (s[n // 2 - 1] + s[n // 2]) / 2.0


def library_versions(x: Sequence[float]):
    """numpy / scipy implementations of the same measures, for cross-checking."""
    from scipy import stats

    arr = np.asarray(x, dtype=float)
    return {
        "range (np.ptp)": float(np.ptp(arr)),
        "variance ddof=1 (np.var)": float(np.var(arr, ddof=1)),
        "std ddof=1 (np.std)": float(np.std(arr, ddof=1)),
        "IQR (scipy.stats.iqr)": float(stats.iqr(arr)),
        "CV (scipy.stats.variation, ddof=1)": float(stats.variation(arr, ddof=1)),
        "MAD scaled (scipy.stats.median_abs_deviation)": float(
            stats.median_abs_deviation(arr, scale="normal")
        ),
    }


if __name__ == "__main__":
    data = [4, 8, 6, 5, 3, 9, 7, 11, 6, 100]
    print("data:", data, "\n")
    print("--- from scratch ---")
    print("range            :", value_range(data))
    print("variance (n-1)   :", variance(data))
    print("std dev  (n-1)   :", stddev(data))
    print("IQR              :", iqr(data))
    print("CV               :", coefficient_of_variation(data))
    print("MAD about mean   :", mean_abs_deviation(data, "mean"))
    print("MAD about median :", mean_abs_deviation(data, "median"))
    print("median abs dev   :", median_abs_deviation(data))
    print("\n--- library ---")
    for k, v in library_versions(data).items():
        print(f"{k:46s}: {v}")

File: techniques/dispersion/r/dispersion.R
# Measures of dispersion / spread (Reference §1.2)
# From-scratch base-R implementations plus the idiomatic base/stats calls.
# Run with:  Rscript dispersion.R
#
# Inputs used below:
#   x      : numeric vector (the sample)
#   ddof   : "delta degrees of freedom" -- divisor is n - ddof
#            (ddof = 1 for the sample variance, the default; ddof = 0 for the population)
#   p      : probability in [0, 1] (for quantile_type7)
#   center : "mean" or "median" (which center to use for mean_abs_deviation)
#   scale  : MAD multiplier; 1.4826 gives a consistent estimate of sigma at the normal

value_range <- function(x) max(x) - min(x)

variance_scratch <- function(x, ddof = 1) {
  n <- length(x); stopifnot(n - ddof > 0)
  m <- mean(x); sum((x - m)^2) / (n - ddof)
}

sd_scratch <- function(x, ddof = 1) sqrt(variance_scratch(x, ddof))

quantile_type7 <- function(x, p) {
  s <- sort(x); n <- length(s)
  if (n == 1) return(s[1])
  h <- (n - 1) * p; lo <- floor(h); frac <- h - lo
  hi <- min(lo + 1, n - 1)
  s[lo + 1] + frac * (s[hi + 1] - s[lo + 1])
}

iqr_scratch <- function(x) quantile_type7(x, 0.75) - quantile_type7(x, 0.25)

cv_scratch <- function(x, ddof = 1) {
  m <- mean(x); stopifnot(m != 0); sd_scratch(x, ddof) / m
}

mean_abs_deviation <- function(x, center = c("mean", "median")) {
  center <- match.arg(center)
  c0 <- if (center == "mean") mean(x) else median(x)
  mean(abs(x - c0))
}

median_abs_deviation_scratch <- function(x, scale = 1.4826) {
  med <- median(x); median(abs(x - med)) * scale
}

# Library equivalents: var, sd, range/diff, IQR, stats::mad
library_versions <- function(x) list(
  `range (base)`        = diff(range(x)),
  `var (base, n-1)`     = var(x),
  `sd (base, n-1)`      = sd(x),
  `IQR (base)`          = IQR(x),
  `mad scaled (stats)`  = mad(x),                 # default constant = 1.4826
  `CV (sd/mean)`        = sd(x) / mean(x)
)

if (sys.nframe() == 0) {
  data <- c(4, 8, 6, 5, 3, 9, 7, 11, 6, 100)
  cat("data:", data, "\n\n--- from scratch ---\n")
  cat("range            :", value_range(data), "\n")
  cat("variance (n-1)   :", variance_scratch(data), "\n")
  cat("std dev  (n-1)   :", sd_scratch(data), "\n")
  cat("IQR              :", iqr_scratch(data), "\n")
  cat("CV               :", cv_scratch(data), "\n")
  cat("MAD about mean   :", mean_abs_deviation(data, "mean"), "\n")
  cat("MAD about median :", mean_abs_deviation(data, "median"), "\n")
  cat("median abs dev   :", median_abs_deviation_scratch(data), "\n\n--- library ---\n")
  lv <- library_versions(data)
  for (nm in names(lv)) cat(sprintf("%-22s: %s\n", nm, lv[[nm]]))
}
